### PR TITLE
Easter cleanup additions

### PR DESCRIPTION
**Pull request by philipppahl**
## 

**Changes proposed in this pull request:**
- Introduction of base class JSONSerializable for specs

**Todos**
- [x] Unit Tests
- [x] Integration Tests
- [x] Examples

/cc @amatthies

### CODE REVIEW

**Comment by amatthies** [comment]
-1 for making `JSONSerializable` the base class of all floto spec classes, imho that's too much inheritance just for code reuse. _Favor object composition over class inheritance_

**Comment by philipppahl** [comment]
@amatthies 
Static functions are now in floto.specs.serializer and used by specs. Removed JSONSerializable.

**Comment by amatthies** [comment]
@philipppahl Much better. Other than that:
- +1 for  removing `_deserialized` from `DeciderSpec`. Please do the same for the others.
- -1 for the method name `copy_args_wo_type`.

**Comment by philipppahl** [comment]
@amatthies 
My idea was to have a generic deserialized in the base classes and "implementations" (_deserialized) in the child classes. I'll remove it, so that it's consistent. We then have to check the type elsewhere in the code at the deserialization step.

**Comment by philipppahl** [comment]
@amatthies 
- removed _deserialized
- ActivityTask and specs.task.ChildWorkflow have id_ parameter for consistency
- requires parameter is now list of ids in favor of list of tasks: Better solution as we do not have to serialize the complete tasks for the requires parameter and it contains all information which is needed.

ExecutionGraph:
- Only handles the requirements, does not use tasks any more (better separation of concerns, more transparent). Cleaner solution wrt. list of ids in `requires` of tasks
- Removed transitive reduction O(n^2.4). Can become computing intense in case of many generators and many tasks. Was introduced to improve fewer queries to the history ("isTaskCompleted"). Instead it is cheaper to generate the list of completed tasks and query the list.
- Checks for cycles in graph

**Comment by amatthies** [comment]
+1

File: floto/specs/serializer.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    cpy['type'] = class_path
+    return dict_to_ordered_dict(cpy)
+
+def get_class(class_path):
+    parts = class_path.split('.')
+    module_name = '.'.join(parts[:-1])
+    class_name = parts[-1]
+    return getattr(sys.modules[module_name], class_name)
+
+def copy_dict(dict_, filter_keys=[]):
```
Comment (amatthies) [inline floto/specs/serializer.py]: default argument is mutable. Don't do this.

**Pull request merged**